### PR TITLE
Fix new `size_of` for IR.

### LINKED ISSUE #717: Strings are only half implemented when using IR.
Currently the string implementation is half-arsed, which is probably worse than not done at all.  I think it was waiting for static spans or something, but should be fixed ASAP.

### PR DESCRIPTION
**Pull request by otrho**
Also fix compiling strings for IR.  Blocked by #852, need that to be merged first since this PR is attached to it.

Closes #717.

### CODE REVIEW

**Comment by otrho** [comment]
All of the stuff at the top of this PR, the `type_analyzer` and aggregate analysis is just moving what I already had in `asm_generation/from_ir.rs` to calculate the size of types over to `optimize.rs`.   Then it can be used for the new `SizeOf` node compilation.

File: sway-core/src/asm_generation/from_ir.rs [hunk 1]
```diff
@@ -1639,16 +1654,70 @@ impl<'ir> AsmBuilder<'ir> {
...
+        ConstantValue::Undef => unreachable!("Cannot convert 'undef' to a literal."),
+        ConstantValue::Unit => Literal::U64(0), // No unit.
+        ConstantValue::Bool(b) => Literal::Boolean(*b),
+        ConstantValue::Uint(n) => Literal::U64(*n),
+        ConstantValue::B256(bs) => Literal::B256(*bs),
+        ConstantValue::String(str) => Literal::String(crate::span::Span {
+            span: pest::Span::new(std::sync::Arc::from(str.as_str()), 0, str.len()).unwrap(),
+            path: None,
+        }),
+        ConstantValue::Array(_) => unimplemented!(),
```
Comment (emilyaherbert) [inline sway-core/src/asm_generation/from_ir.rs:1670]: Can we remove `unimplemented!()` and instead give back an `Internal` compiler error message?

File: sway-core/src/asm_generation/from_ir.rs [hunk 2]
```diff
@@ -1639,16 +1654,70 @@ impl<'ir> AsmBuilder<'ir> {
...
+        ConstantValue::Struct(_) => unimplemented!(),
```
Comment (emilyaherbert) [inline sway-core/src/asm_generation/from_ir.rs:1671]: Remove `unimplemented!()`?

File: sway-core/src/asm_generation/from_ir.rs [hunk 3]
```diff
@@ -1639,16 +1654,70 @@ impl<'ir> AsmBuilder<'ir> {
...
+            Type::Union(aggregate) => {
+                self.analyze_aggregate(context, aggregate);
+                self.aggregate_max_field_size(context, aggregate)
+            }
+            Type::ContractCaller(_) => {
+                // We only store the address.
+                32
+            }
+            Type::Contract => {
+                unimplemented!("do contract/contract caller have/need a size?")
```
Comment (emilyaherbert) [inline sway-core/src/asm_generation/from_ir.rs:1707]: Remove `unimplemented!()`?

**Review by emilyaherbert — CHANGES_REQUESTED** [review]
LGTM---just left a few comments.

File: sway-core/src/asm_generation/from_ir.rs (hunk 1 above)
Comment (otrho) [reply sway-core/src/asm_generation/from_ir.rs]: I started to get this to return an error and gave up after a bit.  The diff to get an error from this deep would probably be several times larger than this PR is now.

The thing is, this is semi-temporary code.  While we have the hybrid old-ASMgen-with-optional-IR-ASMgen we're still sharing certain bits of code, including the 'data section' stuff.  The only reason we need to convert an IR type to an AST literal is to put it in the data section which is then used in ASMgen to line up consts at the end of the code and store a pointer to them.

Once we switch to IR by default and remove the old ASMgen, this won't be needed any more, there will be no need for this function.  I have a note about it in #738.

File: sway-core/src/asm_generation/from_ir.rs (hunk 2 above)
Comment (otrho) [reply sway-core/src/asm_generation/from_ir.rs]: Ditto.

File: sway-core/src/asm_generation/from_ir.rs (hunk 3 above)
Comment (otrho) [reply sway-core/src/asm_generation/from_ir.rs]: `Contract` isn't a real type, it's a bit weird.  I've given it a size of 0.

**Review by sezna — APPROVED** [review]
utACK

**Comment by adlerjohn** [comment]
Pinging @emilyaherbert for re-review

**Review by emilyaherbert — APPROVED** [review]

**Pull request merged**